File: src/main/java/toker/warbandscripts/panel/service/ServerService.java
package toker.warbandscripts.panel.service;

import org.springframework.data.crossstore.ChangeSetPersister;
import org.springframework.stereotype.Service;
import toker.warbandscripts.panel.entity.Server;
import toker.warbandscripts.panel.entity.ServerConfiguration;
import toker.warbandscripts.panel.repository.ServerConfigurationRepository;
import toker.warbandscripts.panel.repository.ServerRepository;

import javax.persistence.criteria.JoinType;
import java.io.File;
import java.nio.file.Files;
import java.util.Collection;
import java.util.Optional;

@Service
public class ServerService {

    private ServerRepository serverRepository;
    private ServerConfigurationRepository serverConfigurationRepository;

    public ServerService(ServerRepository serverRepository,
                         ServerConfigurationRepository serverConfigurationRepository) {
        this.serverRepository = serverRepository;
        this.serverConfigurationRepository = serverConfigurationRepository;
    }

    public Server getServer(Integer serverId, String... with) throws ChangeSetPersister.NotFoundException {
        return serverRepository.findOne((root, query, builder) -> {
            for (String attr : with) {
                root.fetch(attr, JoinType.LEFT);
            }
            return builder.equal(root.get("id"), serverId);
        }).orElseThrow(ChangeSetPersister.NotFoundException::new);
    }

    public Server getServerByKey(String key) {
        return serverRepository.findByKey(key).orElse(null);
    }

    public Optional<ServerConfiguration> getServerConfiguration(int serverId, String name) {
        return serverConfigurationRepository.findByServerIdAndName(serverId, name);
    }

    public Collection<ServerConfiguration> getServerConfigurations(int serverId) {
        return serverConfigurationRepository.findAllByServerId(serverId);
    }

    public File getMapDir(Server server) {
        File exe = new File(server.getExePath());
        File serverDir = new File(exe.getParent());
        File modulesDir = new File(serverDir, "Modules");
        File moduleDir = new File(modulesDir, server.getModuleName());
        return new File(moduleDir, "SceneObj");
    }
}

File: src/main/java/toker/panel/service/ServerService.kt
package toker.panel.service

import org.springframework.data.crossstore.ChangeSetPersister
import org.springframework.stereotype.Service
import toker.panel.entity.*
import toker.panel.repository.FactionRepository
import toker.panel.repository.ServerConfigurationRepository
import toker.panel.repository.ServerRepository
import toker.panel.repository.ServerStartupCommandRepository
import java.io.File
import java.util.*
import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.JoinType
import javax.persistence.criteria.Root

@Service
class ServerService(private val serverRepository: ServerRepository,
                    private val serverConfigurationRepository: ServerConfigurationRepository,
                    private val serverStartupCommandRepository: ServerStartupCommandRepository,
                    private val factionRepository: FactionRepository) {

    @Throws(ChangeSetPersister.NotFoundException::class)
    fun getServer(serverId: Int?, vararg with: String?): Server {
        return serverRepository.findOne { root: Root<Server?>, _, builder: CriteriaBuilder ->
            for (attr in with) {
                root.fetch<Any, Any>(attr, JoinType.LEFT)
            }
            builder.equal(root.get<Any>("id"), serverId)
        }.orElseThrow { ChangeSetPersister.NotFoundException() }
    }

    fun getServerByKey(key: String): Server {
        return serverRepository.findByKey(key).orElse(null)
    }

    fun getServerConfiguration(serverId: Int, name: String): Optional<ServerConfiguration> {
        return serverConfigurationRepository.findByServerIdAndName(serverId, name)
    }

    fun getServerConfigurations(serverId: Int): Collection<ServerConfiguration> {
        return serverConfigurationRepository.findAllByServerId(serverId)
    }

    fun getModuleDir(server: Server): File {
        val exe = File(server.exePath!!)
        val serverDir = File(exe.parent)
        val modulesDir = File(serverDir, "Modules")
        return File(modulesDir, server.moduleName!!)
    }

    fun getMapDir(server: Server): File {
        return File(getModuleDir(server), "SceneObj")
    }

    fun getStartupCommands(serverId: Int): List<ServerStartupCommand> {
        return serverStartupCommandRepository.findAll { root, _, builder ->
            builder.equal(root.get(ServerStartupCommand_.server).get(Server_.id), serverId)
        }
    }

    fun getFactions(serverId: Int): List<Faction> = factionRepository.findAllByServerId(serverId)
}